### PR TITLE
Jenkins 34825

### PR DESCRIPTION
**Pull request by p4paul**

### CODE REVIEW

**Comment by p4paul** [comment]
The tests seem to fail to find the credential...  
findCredential(String id, Run run) is called, but returns null.  The tests create their own credentials in DefsultEnvironment.createCredentials(...)

**Comment by Dohbedoh** [comment]
Quick reason for all this changes. The P4 credentials were lookup up using the context _Jenkins.getInstance()_ in the [ConnectionHelper|https://github.com/jenkinsci/p4-plugin/blob/master/src/main/java/org/jenkinsci/plugins/p4/client/ConnectionHelper.java#L534-L549] as well as in the [P4CredentialsImpl|https://github.com/jenkinsci/p4-plugin/blob/master/src/main/java/org/jenkinsci/plugins/p4/credentials/P4CredentialsImpl.java#L24]. Those methods are indirectly called by the scm, builders and publishers. The context needs to be passed from these callers.

Furthermore, the credentials used in test methods had the scope SYSTEM. Which works if jobs tested are created at root level. But this is wrong since credentials used in jobs should have the scope GLOBAL.

Also noticed that the scope was not configurable from the UI. _config.jelly_ needed to be renamed to _credentials.jelly_. In short, it was not possible to create SSYTEM scoped P4 credentials from the UI.

Still need to investigate on the build failure.

@reviewbybees

**Comment by p4paul** [comment]
Looks like they are already here. GitHub makes no sense...

**Comment by jglick** [comment]
> scope SYSTEM. Which works if jobs tested are created at root level

Not even then. `SYSTEM` is for `Jenkins` itself, not `Job`s IIUC.

File: src/main/java/org/jenkinsci/plugins/p4/PerforceScm.java [hunk 1]
```diff
@@ -220,8 +224,22 @@ public String getKey() {
 			logger.fine("No credential for perforce");
 			return null;
 		}
+
+		// Retrieve item from request
+		StaplerRequest req = Stapler.getCurrentRequest();
+		Job job = req == null ? null : req.findAncestorObject(Job.class);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/p4/PerforceScm.java]: This smells wrong. `guessBrowser` is not necessarily called from an HTTP handler thread.

Comment (Dohbedoh) [inline src/main/java/org/jenkinsci/plugins/p4/PerforceScm.java]: @jglick Thanks for the feedback. I cannot find another way to get the context (_Item_ or _ItemGroup_ ) from this method. What do you suggest ?

**Comment by p4paul** [comment]
Any progress on `guessBrowser`?  If the `findCredential` method were only to use `Jenkins.getActiveInstance()` would that prevent Folder Domain Credentials from being found? 

The `Jenkins.getActiveInstance()` calls to `findCredential(String credentialsId, ItemGroup context)` which looking at the `CredentialsProvider` seems to look up all Items anyway.

**Comment by p4paul** [comment]
The test only touch `ConnectionHelper.findCredential(... Item item)`.  So out of interest I replaced the method to use Jenkins.getActiveInstance().getAllItems() and iterate over them and all the tests still pass? Are the tests bad or do we not need the item?

	public static P4BaseCredentials findCredential(String credentialsId, Item i) {
		if (credentialsId == null) {
			return null;
		}

		List<Item> items = Jenkins.getActiveInstance().getAllItems();
		for (Item item : items) {
			P4BaseCredentials credentials = CredentialsMatchers.firstOrNull(
					CredentialsProvider.lookupCredentials(P4BaseCredentials.class, item,
							ACL.SYSTEM, Collections.<DomainRequirement>emptyList()),
					CredentialsMatchers.allOf(
							CredentialsMatchers.withId(credentialsId),
							CredentialsMatchers.instanceOf(P4BaseCredentials.class)));
			if (credentials != null) {
				CredentialsProvider.track(item, credentials);
				return credentials;
			}
		}
		return null;
	}

**Comment by Dohbedoh** [comment]
> If the findCredential method were only to use Jenkins.getActiveInstance() would that prevent Folder Domain Credentials from being found?

That's correct. From the job perspective, only credentials defined at the job level or above (parent folder up until the root) are visible. This is expected and guarantee certain level of security in Jenkins as you don't want a job to be able to see/use credentials defined in a different folder branch. The problem here is that I am not sure how to retrieve the context from the `guessBrowser` method. The stapler request is not the best solution according to @jglick 

> So out of interest I replaced the method to use Jenkins.getActiveInstance().getAllItems() and iterate over them and all the tests still pass?

Doing this, the method would be able to see/use credentials defined in all Jenkins items. This is rather insecure as explained above. Also this would probably blow up performances especially if you have a large instance with thousands of items.

I believe that calling the `findCredentials(credentialsId, ItemGroup)` for Jenkins root level would be a suitable solution, like it is done in case we cannot find the context: https://github.com/Dohbedoh/p4-plugin/blob/9041d1d8085a7280ebecf57208d6b42e7068cb4c/src/main/java/org/jenkinsci/plugins/p4/PerforceScm.java#L234
Which is the current behavior of the plugin.

In any case, I think it is worth to get feedback from @jglick or even @stephenc 

Note: I also need to rebase to resolve conflict with master.

**Comment by p4paul** [comment]
Thank you for the explanation, I had a feeling that the Job would be necessary to find the credential.   Sorry about the merge conflict, I had been working with @stephenc to support the latest updates to scm-api for multi-branch.
I guess `Stapler.getCurrentRequest()` is the only way to find the context, Job and then the Credential.

File: src/main/java/org/jenkinsci/plugins/p4/PerforceScm.java (hunk 1 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/p4/PerforceScm.java]: I have no suggestion. `guessBrowser` is not offered any context. What are you trying to solve here?

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/p4/PerforceScm.java]: And why are you doing this calculation now? Do you actually need to contact the server in order to determine whether to use `SwarmBrowser` (whatever that is) or something else? (And what would that something else be?)

`guessBrowser` is just intended as a heuristic, for SCMs which multiple kinds of web-based browsers depending on hosting service, etc. That is probably not the case for P4.

Comment (Dohbedoh) [inline src/main/java/org/jenkinsci/plugins/p4/PerforceScm.java]: > And why are you doing this calculation now? Do you actually need to contact the server in order to determine whether to use SwarmBrowser (whatever that is) or something else?

Apparently, the Swarm URL is a server property `P4.Swarm.URL` so you need to connect to the server in order to retrieve it. Also the server URI is stored in P4 Credentials.

The current implementation for `guessBrowser` is to lookup the credentials specified for this SCM at Jenkins root. Given the context that Jenkins provides. it seems that P4 can simply not properly guess the browser implementation and simply return `null`. Is that correct ?

**Comment by p4paul** [comment]
Thank you - all looks good.  I added a small change to remove the dependency on a running P4D.

**Comment by jglick** [comment]
> the Swarm URL is a server property P4.Swarm.URL so you need to connect to the server in order to retrieve it

This is a problem—this kind of network connection is inappropriate during `guessBrowser`, which is only supposed to inspect static configuration.

Really the work should be pushed back into `getChangeSetLink`, if and when that is called (i.e., the changelog page is actually displayed for this build), caching the Swarm URL transiently if necessary. The `ChangeLogSet.Entry` has a `parent` which in turn has a `run` which you should be able to use to look up credentials.

**Pull request merged**